src/structure/structure.ts

```diff
@@ -267,4 +357,258 @@ export class StructureViewer {
+        cell.appendChild(duplicate);
+
+        this._root.appendChild(cell);
+        } else {
+        cell.style.gridColumn = `${colNum}`;
+        cell.style.gridRow = `${rowNum}`;
+      }
+
+    }
+    /*
+     * Function to initialize the grid instance for nwidgets cells and place
+     * onto the DOM element mapped to `id`
+     */
+    private _setupGrid() {
+
+      // Determine best arrangement for nwidgets
+      const arrangement = bestArrangement(this._nwidgets);
+
+      if (this._widgets.length > this._nwidgets) {
+        sendWarning(`Warning: Eliminating last ${this._widgets.length - this._nwidgets} widgets.`);
+        const wl = this._widgets.length;
+        for (let i = this._nwidgets; i < wl; i++) {
+          this._removeWidget(this._guids[i]);
+        }
+      }
+
+      const root = document.getElementById(this._id);
+      if (root === null) {
+          throw Error(`Could not get the '#${this._id}' element`);
+      }
+
+      // Initialize the grid root if necessary
+      if (this._root === undefined) {
+        this._root = document.createElement('div');
+        this._root.id = 'grid-root';
+        this._root.className = 'chsp-widget-grid';
+
+        // Initializes two narrow rows for the `increase` and `decrease` buttons,
+        // then auto scales the remaining rows
+        root.appendChild(this._root);
+      }
+
+      // Start at the third row to skip the inc/dec buttons
+      let rowNum = 1;
+      let colNum = 1;
+
+      for (let c = 0; c < this._nwidgets; c++) {
+        this._setupCell(c, colNum, rowNum);
+
+        colNum++;
+        if (colNum > arrangement.columns) {
+            rowNum++;
+            colNum = 1;
+          }
+
+        // add a new widget if necessary
+        if (c >= this._widgets.length) {
+          this._widgets.push(new JSmolWidget(`gi-${this._guids[c]}`,
+                                             this._j2spath,
+                                             'chsp-' + this._guids[c]));
+          this._currentList.push({atom: -1, structure: -1, environment: -1});
+          const index = {atom: 0, structure: 0, environment: 0};
+          this.show(index, this._guids[c]);
+          this.onselect(index, this._guids[c]);
+        } else {
+          this._widgets[c].script('refresh');
+        }
+      }
+      return this._root;
+    }
+
+    private _removeWidget(guid: string, force: boolean = false) {
+      //
+      if ( this._widgets.length > 1 || force === true) {
+        const widgetRoot = document.getElementById(guid);
+
+        const gi = this._guids.indexOf(guid);
+        if ( gi >= 0 ) {
+          this.onselect(this._currentList[gi], `${guid}-nuke`);
+        }
+
+        if ( widgetRoot !== null) {
+            const oldRoot = widgetRoot.parentNode;
+            if ( oldRoot !== null) {oldRoot.removeChild(widgetRoot); }
+        }
+        const deadCell = document.getElementById(`gi-${guid}`);
```

Comment:
Most of the uses of `document.getElementById` should be `getById` instead to catch the case where the element id do not exists in the document